src/index.js

```diff
@@ -0,0 +1,756 @@
+       }
+
+       gossipSubPeers.forEach((peer) => {
+           this.log("JOIN: Add mesh link to %s in %s", peer.info.id.toB58String, topic)
+	   this._sendGraft(peer, topic)
+	   peer.topics.add(topic)
+       })
+
+       
+   }
+
+   /**
+    * Leaves a topic
+    * @param {String}
+    *
+    */
+   unsubscribe(topic) {
+       let ok = this.mesh.has(topic)
+       let gmap = this.mesh.get(topic)
+       if (!ok) {
+           return
+       }
+
+       this.log("LEAVE %s", topic)
+
+       this.mesh.delete(topic)
+
+       for (let peer of gmap) {
+           this.log("LEAVE: Remove mesh link to %s in %s", peer.info.id.toB58String, topic)
+	   this._sendPrune(peer, topic)
+	   this.peer.topics.delete(topic)
+	   
+       }
+
+   }
+
+   /**
+    *
+    * @param {Peer}
+    * @param {any}
+    *
+    */
+   publish(from, msg) {
+       this.messageCache.Put(msg)
+
+       // @type Set<string>
+       let tosend = new Set()
+       msg.topicIDs.forEach((topic) => {
+           if (!this.topics.has(topic)) {
+	       continue
+	   }
+
+	   let peersInTopic = this.topics.get(topic)
+	   
+	   // floodsub peers
+	   peersInTopic.forEach((peer) => {
+	       if (peer.info.protocols.has(FloodSubID)) {
+	           tosend.add(peer)
+	       }
+	   })
+
+	   // Gossipsub peers handling
+	   if (!this.mesh.has(topic)) {
+	       // We are not in the mesh for topic, use fanout peers
+	       if (!this.fanout.has(topic)) {
+	           // If we are not in the fanout, then pick any peers
+		   let peers = this._getPeers(topic, GossipSubD)
+
+		   if(peers.size > 0) {
+		       this.fanout.set(topic, peers)
+		   }
+	       }
+	       // Store the latest publishing time
+	       this.lastpub.set(topic, _nowInNano())
+	   }
+
+	   let meshPeers = this.mesh.get(topic)
+	   meshPeers.forEach((peer) => {
+	       tosend.add(peer)
+	   })
+       })
+       // Publish messages to peers
+       tosend.forEach((peer) => {
+           let peerId = peer.info.id.getB58Str()
+	       if (peerId === from || peerId === msg.from) {
+	       continue
+	   }
+	   peer.sendMessages(msg)
+       })
+
+
+   }
+
+   /**
+    * Sends a GRAFT message to a peer
+    *
+    * @param {Peer} peer 
+    * @param {String} topic
+    */
+   _sendGraft(peer, topic) {
+       let graft = [{
+           topicID: topic
+       }]
+
+       let out = this._rpcWithControl(null, null, null, graft, null)
+       if(peer && peer.isWritable()) {
+           peer.write(out)
+	   peer.sendSubscriptions([topic])
+       }
+   }
+
+   /**
+    * Sends a PRUNE message to a peer
+    *
+    * @param {Peer} peer
+    * @param {String} topic
+    *
+    */
+   _sendPrune(peer, topic) {
+       let prune = [RPC.ControlPrune.encode({
+           topicID: topic
+       })]
+
+       let out = _rpcWithControl(null, null, null, null, prune)
+       if(peer && peer.isWritable()) {
+          peer.write(out)
+	  peer.sendUnsubscriptions([topic])
+       }
+
+   }
+
+   _heartbeatTimer() {
```

Comment:
This function needs cleanup, appears to have duplicate code, currently causing errors trying to create a new GossipSub object.

Reply:
Currently, this function doesn't even work. It needs serious debugging. I plan on working on this throughout the week while I write tests.

Reply:
:+1: